lib/client/api_matrix.es6.js

```diff
@@ -72,37 +72,40 @@ foam.CLASS({
             This structure is easy for displaying a nested table.`,
       },
       code: function(browserKeys) {
-        return this.browserApiDAO
-          .where(this.mlang
-            .IN(this.BrowserWebInterfaceJunction.SOURCE_ID, browserKeys))
-          .orderBy(this.BrowserWebInterfaceJunction.TARGET_ID)
-          .select(this.mlang
-            .GROUP_BY(this.BrowserWebInterfaceJunction.TARGET_ID,
-            this.ArraySink.create()))
-          .then((groups) => {
-            let apiMatrix = {};
-            let interfaceFindPromises = [];
-            for (let i = 0; i < groups.groupKeys.length; i++) {
-              let interfaceKey = groups.groupKeys[i];
-              interfaceFindPromises.push(this.interfaceDAO
-                .find(interfaceKey)
-                .then((webInterface) => {
-                  if (!apiMatrix.hasOwnProperty(webInterface.interfaceName)) {
-                    apiMatrix[webInterface.interfaceName] = {};
-                  }
-                  apiMatrix[webInterface.interfaceName][
-                    webInterface.apiName] = {};
-                  let browserApis = groups.groups[interfaceKey].a;
-                  for (let j = 0; j < browserApis.length; j++) {
-                    apiMatrix[webInterface.interfaceName][webInterface
-                      .apiName][browserApis[j].sourceId] = true;
-                  }
-              }));
-            }
-            return Promise.all(interfaceFindPromises).then(() => {
-              return apiMatrix;
-            });
-          });
+        let matrix = {};
+        return Promise.all(browserKeys.map(
+          (browserId) => this.browserDAO.find(browserId)
+            .then((browser) => {
+              // Reject if this browser does not exit in database.
+              if (browser === null) {
+                return Promise.reject(Error(`${browserId} does not exist.`));
+              }
+              // TODO(markdittmer): ContextualizingDAO should take care of this.
+              return browser.cls_.create(browser, this.__subContext__)
+                  .interfaces.select({
+                    put: (iface) => {
+                      let k0 = iface.interfaceName;
+                      let k1 = iface.apiName;
+                      let k2 = browserId;
+                      // Function and Object's built-in methods can be
+                      // interface name or api name.
+                      // So matrix[k0] is not sufficient to check if
+                      // k0 exits in matrix, use hasOwnProperty instead.
+                      if (!matrix.hasOwnProperty(k0)) {
+                        matrix[k0] = {};
+                      }
+                      // hasOwnProperty could also be a apiName,
+                      // thus use another Object's hasOwnProperty
+                      // and call it with 'this' set to foo
+                      if (!({}).hasOwnProperty.call(matrix[k0], k1)) {
```

Comment:
Let's use `Object.prototype.hasOwnProperty.call` to avoid creating a garbage object every time.